pom.xml

```diff
@@ -760,6 +760,12 @@
         <version>1.10.19</version>
         <scope>test</scope>
       </dependency>
+      <dependency>
```

Comment:
Why adding this to the top level pom?

Reply:
We always add to the top level and then in the lower level poms, we reference the dependent modules without listing their versions.

Reply:
also add dependency should have its LICENSE added under /license

Reply:
Is it necessary to include the license file if the dependent project (RxJava) is also licensed under Apache 2.0?

Reply:
I don't think this should cause rxjava to be pulled in from spark core. But the library will be put into the Spark distribution, and anyone depending on both Kubernetes integration and RxJava will have to resolve potential conflicts.

Reply:
@felixcheung (sorry for the delayed answer), @mccheah  a few thoughts/observations:
a) If not mistaken Java 9 has introduced the related interfaces, so rx-java might not be needed in the future.
b) Some more scala centric implementations:  [akka streams](https://doc.akka.io/docs/akka/current/stream/stream-integrations.html#integrating-with-reactive-streams) instead, a Reactive Streams and JDK 9+ java.util.concurrent.Flow-compliant implementation. This does not depend on rx-java.
Also there is [monix] (https://github.com/monix/monix), and (RxScala)[https://github.com/ReactiveX/RxScala] seems outdated. A comparison between monix and others (here)[https://monix.io/docs/2x/reactive/observable-comparisons.html].
One question is how about performance at scale when you get events from hundreds of executors at once which framework should work best? Should we worry about this? 
Rx-java is light but also monix might be light at the end of the day or akka streams, although its purpose goes beyond observables etc.
c) I built the distro from the PR, so rxjava-2.1.13.jar, reactive-streams-1.0.2.jar are added to the jars. Spark core does not seem to have any classes in it.

Reply:
> f not mistaken Java 9 has introduced the related interfaces, so rx-java might not be needed in the future when Spark will update the supported java version.

Spark will need to support Java 8 for the foreseeable future. We'll be using a Java 8 only compatible solution for awhile.

> Some more scala centric implementations: akka streams a Reactive Streams and JDK 9+ java.util.concurrent.Flow-compliant implementation. This does not depend on rx-java.
Also there is monix, and RxScala seems outdated. A comparison between monix and others here.

Does Akka streams without JDK 9 depend on ReactiveX? If so then we have the same dependency problem. As for monix vs. Akka vs. RxJava, think that's an implementation detail - at the end of the day the main concern is the dependency addition and either way we're adding an external dependency vs. implementing the semantics ourselves. If we think monix / Akka is the more elegant solution over RxJava we can switch to that, but given how little we're actually doing that's actually in the functional programming style I don't think the difference between Java and Scala is significant in this particular instance.

> From the dependencies already brought in by Spark is there a lib that could provide similar functionality, if not based on Reactive stuff some other way queues or something?

I didn't see any but I am open to being corrected here. Spark implements an event queue but it doesn't suffice for the paradigm of having multiple subscribers process snapshots at different intervals.

Reply:
> One question is how about performance at scale when you get events from hundreds of executors at once which framework should work best? Should we worry about this?

One already runs into this problem since we open a Watch that streams all events down anyways. In any implementation where we want events to be processed at different intervals, there needs to be some buffering or else we choose to ignore some events and only look at the most up to date snapshot at the given intervals. As discussed in https://github.com/apache/spark/pull/21366#discussion_r194181797 we really want to process as many events as we get as possible, so we're stuck with buffering somewhere, and regardless of the observables or reactive programming framework we pick we still have to effectively store `O(E)` items, `E` being the number of events. And aside from the buffering we'd need to consider the scale of the stream of events flowing from the persistent HTTP connection backing the Watch. In this regard we are no different from the other custom controllers in the Kubernetes ecosystem which have to handle managing large number of pods.

Reply:
> Does Akka streams without JDK 9 depend on ReactiveX?
It depends on reactive streams library so you dont need to bring rx-Java in. @ktoso correct me if I am wrong.

> In this regard we are no different from the other custom controllers in the Kubernetes ecosystem which have to handle managing large number of pods.
Have an example of a specific controller to get a better understanding?

Reply:
Seems some confusion is here?
Akka Streams does not depend on Rx of course, they both alternative implementations of Reactive Streams ( http://reactive-streams.org/ ) which have been included in JDK9 as `java.util.concurrent.Flow.*` and Akka also implements those, but does not require JDK9; you can use JDK8 + RS and if you use JDK9 you could use the JDK's types but it's not required.

Anything else I should clarify or review here? For inter-op purposes it would be good to not expose on a specific implementation but expose the reactive-streams types (`org.reactivestreams.Publisher` etc), but that only matters if the types are exposed. As for including dependencies in core Spark -- I would expect this to carry quite a bit of implications though don't know Spark's rules about it (ofc less dependencies == better for users, since less chances to version-clash with libraries they'd use)

Reply:
These types are not exposed - they're only implementation details in the Kubernetes module. Furthermore the RxJava dependency will be in the Spark distribution but is not a dependency pulled in by spark-core.

It sounds like there is some contention with the extra dependency though, so should we be considering implementing our own mechanisms from the ground up? I think the bottom line question is: can spark-kubernetes, NOT spark-core, pull in RxJava?

Reply:
I ended up just removing reactive programming entirely - the buffering is implemented manually. Please take a look.

Reply:
Thanx @ktoso!